include/zephyr/sys/zync.h

```diff
@@ -0,0 +1,340 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_SYS_ZYNC_H
+#define ZEPHYR_SYS_ZYNC_H
+
+#include <string.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/kernel_structs.h>
+#include <zephyr/syscall.h>
+
+#define K_ZYNC_ATOM_VAL_BITS 24
+#define K_ZYNC_ATOM_VAL_MAX (BIT(K_ZYNC_ATOM_VAL_BITS) - 1)
+
+/** @brief Zephyr atomic synchronization primitive
+ *
+ * The zync atom stores the counted state variable for a struct
+ * k_zync, in such a way that it can be atomically modified.
+ */
+typedef union {
+	atomic_t atomic;
+	struct {
+		uint32_t val : K_ZYNC_ATOM_VAL_BITS; /* Value of the lock/counter */
+		bool waiters : 1;                    /* Threads waiting? */
+		uint32_t ptype : 2;                  /* Pair type */
+	};
+} k_zync_atom_t;
+
+#if defined(CONFIG_ZYNC_PRIO_BOOST) || defined(CONFIG_ZYNC_RECURSIVE) \
+	|| defined(CONFIG_ZYNC_VALIDATE)
+#define Z_ZYNC_OWNER 1
+#endif
+
+struct k_zync_cfg {
+	uint32_t atom_init;
+	IF_ENABLED(CONFIG_ZYNC_MAX_VAL, (uint32_t max_val : K_ZYNC_ATOM_VAL_BITS;))
+	bool fair : 1;
+	IF_ENABLED(CONFIG_ZYNC_PRIO_BOOST, (bool prio_boost : 1;))
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE, (bool recursive : 1;))
+};
+
+/* @brief Fundamental Zephyr thread synchronization primitive
+ *
+ * @see `k_zync()`
+ */
+struct k_zync {
+	struct k_spinlock lock;
+	_wait_q_t waiters;
+	struct k_zync_cfg cfg;
+	IF_ENABLED(Z_ZYNC_OWNER, (struct k_thread *owner;))
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE, (uint32_t rec_count;))
+};
+
+#define Z_ZYNC_MVCLAMP(v) ((v) == 0 ? K_ZYNC_ATOM_VAL_MAX \
+			      : CLAMP((v), 0, K_ZYNC_ATOM_VAL_MAX))
+
+#define K_ZYNC_INITIALIZER(init, isfair, rec, prioboost, maxval) {	\
+	.cfg.atom_init = init,						\
+	IF_ENABLED(CONFIG_ZYNC_MAX_VAL,					\
+		   (.cfg.max_val = Z_ZYNC_MVCLAMP(maxval),))		\
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE,				\
+		   (.cfg.recursive = rec,))				\
+	IF_ENABLED(CONFIG_ZYNC_PRIO_BOOST,				\
+		   (.cfg.prio_boost = prioboost,))			\
+	.cfg.fair = (isfair) }
+
+#define K_ZYNC_DEFINE(name, init, isfair, rec, prioboost, maxval)	\
+	struct k_zync name = K_ZYNC_INITIALIZER(init, isfair, rec, prioboost, maxval);
+
+/** @brief Atomically modify a k_zync_atom
+ *
+ * This macro heads a code block which is responsible for assigning
+ * fields of "new_atom" in terms of the (repeatedly-re-read)
+ * "old_atom", attempting to set the value with `atomic_cas()`, and
+ * repeating the process if the value was changed from another
+ * context.  Code can exit the loop without setting any value by using
+ * a break statement.
+ *
+ * No other modification of the values inside a k_zync_atom from
+ * potentially simultaneous contexts is permitted.
+ *
+ * @note The atom_ptr argument is expanded multiple times in the macro
+ * body and must not have side effects.
+ *
+ * @param atom_ptr Pointer to a k_zync_atom to modify
+ */
+#define K_ZYNC_ATOM_SET(atom_ptr)					           \
```

Comment:
It's worth pointing out that this API looks a little like a spinlock because that's basically what it is.  Which opens the question of whether we should bring spinlocks into the fold and implement them in terms of a k_zync_atom_t, thus unifying ZYNC_VALIDATE and SPIN_VALIDATE.  That seemed a bit too far for me in a first version, but something to consider.